sdk/cosmosdb/cosmos/src/client/SasToken/SasToken.ts

```diff
@@ -0,0 +1,483 @@
+import { DataPlanePermissionScope } from './DataPlanePermissionScope';
+import { ControlPlanePermissionScope } from './ControlPlanePermissionScope';
+import { SasTokenPermissionKind } from './SasTokenPermissionKind';
+import { SasTokenProperties } from './SasTokenProperties';
+import { hmac } from "../../utils/hmac";
+import { CosmosContainerChildResourceKind } from './CosmosContainerChildResourceKind';
+import { CosmosKeyType } from './CosmosKeyType';
+import { Paths } from './Paths';
+import { encodeUTF8, encodeBase64 } from "../../utils/encode";
+
+export class SasToken implements SasTokenProperties {
+  [x: string]: any;
+
+  private AUTH_PREFIX: string = "type=sas&ver=1.0&sig=";
+
+  // private SAS_TOKEN_SEPARATOR: string = ";";
+
+  public user: string;
+  public userTag: string;
+  public databaseName: string;
+  public containerName: string;
+  public resourceName: string;
+  public resourcePath: string;
+  public resourceKind: CosmosContainerChildResourceKind;
+  public partitionKeyValueRanges: SasTokenPartitionKeyValueRange[];
+  public startTime: Date;
+  public expiryTime: Date;
+  public keyType: number;
+  public controlPlaneReaderScope: number;
+  public controlPlaneWriterScope: number;
+  public dataPlaneReaderScope: number;
+  public dataPlaneWriterScope: number;
+  public cosmosContainerChildResourceKind: CosmosContainerChildResourceKind;
+  public cosmosKeyType: CosmosKeyType;
+
+  public constructor() {
+    this.user = "";
+    this.userTag = "";
+    this.databaseName = "";
+    this.containerName = "";
+    this.resourceName = "";
+    this.resourcePath = "";
+    this.startTime.setDate(Date.now());
+    this.expiryTime = null;
+    this.controlPlaneReaderScope = 0;
+    this.controlPlaneWriterScope = 0;
+    this.dataPlaneReaderScope = 0;
+    this.dataPlaneWriterScope = 0;
+    this.keyType = 0;
+  }
+  public create(user: string, databaseName: string, containerName: string): SasTokenProperties {
+    const token = new SasToken()
+    token.setUser(user)
+    token.setDatabaseName(databaseName)
+    token.setContainerName(containerName)
+    return token
+  }
+  public sasTokenValueUsingHMAC(key: string): string;
+  public sasTokenValueUsingHMAC(key: string, keyType?: CosmosKeyType) {
+
+    if (((key == null)
+      || key === "")) {
+      return this.getSasTokenWithHMACSHA256(key);
+    } else if (typeof keyType === "object") {
+      switch (this.keyType) {
+        case CosmosKeyType.PRIMARY_MASTER:
+          this.keyType = 1;
+          break;
+        case CosmosKeyType.SECONDARY_MASTER:
+          this.keyType = 2;
+          break;
+        case CosmosKeyType.PRIMARY_READONLY:
+          this.keyType = 3;
+          break;
+        case CosmosKeyType.SECONDARY_READONLY:
+          this.keyType = 4;
+          break;
+        default:
+          throw new ErrorEvent("keyType");
+          break;
+      }
+      return this.SasTokenWithHMACSHA256(key);
+    }
+    else {
+      throw new ErrorEvent("key");
```

Comment:
```suggestion
      throw new Error("key");
```
Error is what you want here